src/Indigo/Contracts/Holdings/Impl.hs

```diff
@@ -0,0 +1,167 @@
+          )
+          (() <$ failCustom_ #notInTransferAdminRightsMode)
+    , #cIsAdmin //-> \v -> do
+        doc $ DDescription
+          "Check whether address is admin. Returns `True` if the address is an admin."
+        flip view_ v $ \addr -> return $ (storage !. #fields !. #admin) ==. addr
+    , #cTransfer //-> \tParams -> do
+        ensureSenderIsAdmin
+        -- pause check is already done in Lorentz code
+        ensureIsTransferable
+        let from = tParams !. #from
+            to = tParams !. #to
+            constraints = pair (Name #isWhitelisted $ from .: nil)
+                               (Name #isNonBlacklisted $ from .: to .: nil)
+        checkSafelistConstraints constraints
+        transfer tParams
+    , #cApprove //-> \aParams -> do
+        -- pause check is already done in Lorentz code
+        let spender = aParams !. #spender
+            constraints = pair (Name #isWhitelisted $ spender .: nil)
+                               (Name #isNonBlacklisted $ spender .: Sender .: nil)
+        checkSafelistConstraints constraints
+        approve aParams
+    , #cGetAllowance //-> getAllowance
+    , #cGetBalance //-> getBalance
+    , #cGetTotalSupply //-> getTotalSupply
+    , #cMint //-> \mParams -> do
+        ensureNotPaused
+        -- admin check is already done in Lorentz code
+        let to = mParams !. #to
+            constraints = pair (Name #isWhitelisted $ to .: nil)
+                               (Name #isNonBlacklisted $ to .: nil)
+        checkSafelistConstraints constraints
+        mint mParams
+    , #cBurn //-> \bParams -> do
+        ensureNotPaused
+        -- admin check is already done in Lorentz code
+        burn bParams
+    , #cBurnAll //-> \_ -> do
+        doc $ DDescription "Destroy all tokens and allowances."
+        ensureSenderIsAdmin
+        ensureNotPaused
+        setField_ storage #ledger emptyBigMap
+    , #cSetPause //-> \paused -> do
+        -- admin check is already done in Lorentz code
+        setPause paused
+    , #cSetTransferable //-> \transferable -> do
+        doc $ DDescription "Change transferable flag."
+        ensureSenderIsAdmin
+        setField_ storage #fields $
+          (storage !. #fields) ~. (#transferable, transferable)
+    , #cGetTokenMeta //-> \v -> do
+        doc $ DDescription "Get token meta data: name, symbol and id."
+        flip view_ v $ \_ -> return (storage !. #fields !. #tokenMeta)
+    )
+
+checkSafelistConstraints
+  :: ( HasStorage Storage, HasSideEffects
+     , constraints :~> SL.EnsureSafelistConstraintsParam
+     )
+  => constraints -> IndigoProcedure s
+checkSafelistConstraints c = do
+  ifJust (storage !. #fields !. #mbSafelistAddress)
+    (\addr ->
+       ifJust (convertEpAddressToContract @SL.EnsureSafelistConstraintsParam addr)
+         (\contractAddr -> do
+             transferTokens c (toMutez 0) contractAddr
+         )
+         (() <$ failCustom_ #invalidSafelistAddr))
+    (return ())
```

Comment:
Is this correct? In case of no Safelist this should always pass?

Reply:
Yes